### PR TITLE
HDDS-5167. Add clean option for Freon randomkeys to clean generated objects

### PR DESCRIPTION
**Pull request by symious**
## What changes were proposed in this pull request?

Freon randomkeys would generate a lot of random volumes, buckets and keys, it would be quite difficult for users to delete these objects. This patch added a new option named "--clean-objects" to clean all the generated objects.

## What is the link to the Apache JIRA

https://issues.apache.org/jira/browse/HDDS-5167

## How was this patch tested?

Unit test.

### CODE REVIEW

**Review by adoroszlai — COMMENTED** [review]
Thanks @symious for working on this.  I think cleaning up generated objects is a nice feature.

Please note that `RandomKeyGenerator` is obsoleted by a more flexible Freon framework, where each command is responsible for one type of task.  Eg. `OzoneClientKeyGenerator` only generates keys via regular Ozone client.  There is a corresponding `OzoneClientKeyValidator` to validate the same data.

I think it would be nice if the cleanup logic would be added in this framework.  Eg. we could add `OzoneClientKeyRemover` to delete a set of keys.  Benefits:
1. This cleanup could be applied to keys created by other generators (eg. `S3KeyGenerator`).
2. We could add other cleanup implementations for different kinds of objects (`OmBucketGenerator` only creates buckets).

**Comment by symious** [comment]
@adoroszlai Thanks for the review and advice. I will try to generalize the clean feature for the framework.

**Comment by symious** [comment]
@adoroszlai Updated the version to generalize the clean operation, could you help to check?

**Review by adoroszlai — COMMENTED** [review]
Thanks @symious for updating the patch.

I'd like to clarify that my suggestion was to implement clean-up as separate Freon "generators", similar to how key validation is done (see `OzoneClientKeyValidator`).  Some benefits:

1. functionality: Allows measuring the performance of object deletion.
2. flexibility: It allows cleanup after the fact.  With the current implementation I'm out of luck if I forget `--clean-up` when running the generator.
3. design: Less coupling.
4. generator performance: No need to keep track of the names of all objects created by the generator.  With large-scale tests this might be a bottleneck.  Also, no synchronization needed.
5. code: Less duplication eg. related to progress bar and executor.

On the flip side: the objects created during initialization (volume and maybe bucket) need to be deleted separately.  Since these are usually very few in number, I think this is acceptable.  I agree that the original `RandomKeyGenerator` is different in this respect, too.  We can keep the integrated `--clean-up` flag in that class.

Sorry if it was unclear.

**Comment by symious** [comment]
@adoroszlai Thanks for the detailed clarification.
If we separate the class for the delete function, I think one class should work fine for all the key and bucket generators? I didn't do that because I found that class quite dangerous, removing all the keys in a bucket or all buckets in a volume. It's like planting a bomb that can be misused by malicious people or users not familiar with the tool. So I build the clean option inside each generator making sure only the generated objects get cleaned.
It's my personal view, if this is not a problem, then I'll build a separate class for the deleting function.
For the performance monitoring, I think the data are deleted by datanodes asynchronously, seems we can't monitor the deletion precisely?

**Comment by adoroszlai** [comment]
> removing all the keys in a bucket or all buckets in a volume

Freon creates objects with a specific pattern, composed of a prefix and a counter.  The prefix is either random or user-specified.  The validator can only be used with a user-specified prefix, as it has to read previously created keys.  The counter starts at 0 and counts up to N-1, where N is user-specified.

Similarly, the new cleaner would only remove items `$prefix/0`, `$prefix/1`, etc. with the given prefix.  I don't think that would be a dangerous tool.

File: hadoop-ozone/integration-test/src/test/java/org/apache/hadoop/ozone/freon/TestOzoneClientKeyGenerator.java [hunk 1]
```diff
@@ -91,4 +92,41 @@ public void testOzoneClientKeyGenerator() throws Exception {
     shutdown(cluster);
   }
 
+  @Test
+  public void testCleanUpOzoneClientKeyGenerator() throws Exception {
```
Comment (adoroszlai) [inline hadoop-ozone/integration-test/src/test/java/org/apache/hadoop/ozone/freon/TestOzoneClientKeyGenerator.java:96]: Also, please note that this test class is being removed in #2194.  I think similar test case can be added to acceptance tests (`hadoop-ozone/dist/src/main/smoketest/freon`).

File: hadoop-ozone/tools/src/main/java/org/apache/hadoop/ozone/freon/BaseFreonGenerator.java [hunk 2]
```diff
@@ -406,6 +432,24 @@ public void ensureVolumeExists(
...
+   * Create missing target bucket for S3KeyGenerator.
+   */
+  public void ensureBucketExist(AmazonS3 s3, String bucketName) {
+    HeadBucketRequest headBucketRequest = new HeadBucketRequest(bucketName);
+    try {
+      s3.headBucket(headBucketRequest);
+    } catch (AmazonServiceException ex) {
+      if (ex.getErrorCode().equals("NoSuchBucket")) {
+        s3.createBucket(bucketName);
+        setBucketCreated();
```
[the remark this answers is not in the document]
Comment (symious) [reply hadoop-ozone/tools/src/main/java/org/apache/hadoop/ozone/freon/BaseFreonGenerator.java]: Removed S3 code from `baseFreonGenerator`.

File: hadoop-ozone/integration-test/src/test/java/org/apache/hadoop/ozone/freon/TestOzoneClientKeyGenerator.java (hunk 1 above)
Comment (symious) [reply hadoop-ozone/integration-test/src/test/java/org/apache/hadoop/ozone/freon/TestOzoneClientKeyGenerator.java]: Added Robot test for remover.

**Comment by symious** [comment]
@adoroszlai Updated a new version of remover, could you help to check?

**Review by adoroszlai — COMMENTED** [review]
Thanks a lot @symious for updating the patch.  I only have minor comments left.

File: hadoop-ozone/dist/src/main/smoketest/freon/remove.robot [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+    ${result} =        Execute                ozone freon ockg ${OM_HA_PARAM} -t=1 -n=1 -p ocr${PREFIX}
+                       Should contain         ${result}   Successful executions: 1
+
+Ozone Client Remover (KEY)
+    ${result} =        Execute                ozone freon ocr ${OM_HA_PARAM} -t=1 -n=1 -p ocr${PREFIX} --remove-key
+                       Should contain         ${result}   Successful executions: 1
+
+OM Bucket Generator For Remover
+    ${result} =        Execute                ozone freon ombg ${OM_HA_PARAM} -t=1 -n=1 -p ocr${PREFIX}
+                       Should contain         ${result}   Successful executions: 1
```
[the remark this answers is not in the document]
Comment (symious) [reply hadoop-ozone/dist/src/main/smoketest/freon/remove.robot]: Changed to `Setup` in robot script.

File: hadoop-ozone/dist/src/main/smoketest/freon/remove.robot [hunk 4]
```diff
@@ -0,0 +1,43 @@
...
+${PREFIX}    ${EMPTY}
+
+*** Test Cases ***
+Ozone Client Remover (Error)
+    ${result} =        Execute and checkrc    ozone freon ocr ${OM_HA_PARAM} -t=1 -n=1 -p ocr${PREFIX} --remove-key --remove-bucket    255
+                       Should contain         ${result}   Invalid Option
+
+Ozone Client Key Generator For Remover
+    ${result} =        Execute                ozone freon ockg ${OM_HA_PARAM} -t=1 -n=1 -p ocr${PREFIX}
+                       Should contain         ${result}   Successful executions: 1
```
[the remark this answers is not in the document]
Comment (symious) [reply hadoop-ozone/dist/src/main/smoketest/freon/remove.robot]: Changed to `Setup` in robot script.

File: hadoop-ozone/tools/src/main/java/org/apache/hadoop/ozone/freon/OzoneClientRemover.java [hunk 5]
```diff
@@ -0,0 +1,127 @@
...
+      description = "If turned on, Remover will be used to remove buckets. "
+          + "Can only set one option between \"--remove-key\" "
+          + "and \"--remove-bucket\"")
+  private boolean isRemoveBucket;
+
+  @Option(names = {"--remove-key"},
+      description = "If turned on, Remover will be used to remove keys. "
+          + "Can only set one option between \"--remove-key\" "
+          + "and \"--remove-bucket\"")
+  private boolean isRemoveKey;
```
[the remark this answers is not in the document]
Comment (symious) [reply hadoop-ozone/tools/src/main/java/org/apache/hadoop/ozone/freon/OzoneClientRemover.java]: Created separate `OzoneClientKeyRemover` and `OmBucketRemover`.

**Comment by symious** [comment]
@adoroszlai Thanks for the detailed review, have updated the patch to resolve the conversation. Could you help to check?

**Review by adoroszlai — APPROVED** [review]
Thanks @symious for continuously improving the patch.

**Comment by symious** [comment]
@adoroszlai Thanks for the detailed review.

**Pull request merged**